### DISCUSSION TITLE
S3 like storage for files

### DISCUSSION THREAD

**Discussion by arunoruto** [papis/papis#845] [Ideas]
I am getting started with papis, and I love organizing my own "library," but I sometimes want to share documents with others in my group. I would need to store the documents in a central place for this to happen, but a database would not suffice (maybe a NoSQL DB), and filesystems can be cumbersome. Luckily, Amazon had the same problem, and now we have S3 and many alternatives!

I am not saying to have the library only in the database and restructure how Papis works, but what if certain documents/libraries could be synced with an S3 instance like [minio](https://min.io/)? They support user management and are not so hard to set up! And since we are only syncing it, we have offline access to the files.

Since Papis uses the `info.yaml` files for the metadata, they can also be synced with an S3 server. An alternative would be to use [mariaDB](https://mariadb.com/) to store the metadata info and the integrated [S3 storage engine](https://mariadb.com/kb/en/s3-storage-engine/) for the files.

I guess I am not the only one who would appreciate working with Papis in a team, but I am unsure if this approach is reasonable and coincides with the project's philosophy.

**Comment by jghauser** [comment]
If it's just about sharing the documents, then sharing can be done via literally any method. I, for example, use `syncthing` to share my library between my laptop, server, and tablet. In addition, I use a personal Nextcloud instance to share select documents with others.

If what you're interested in is something more like a web front-end that others can use to browse your library (as you mentioned [here](https://github.com/papis/papis/discussions/430#discussioncomment-9346198)), then I think `papis serve` would be what you're after. This command is however fairly new (and I'm not very familiar with it), and I'm not sure which features it has/are missing. It would definitely be very cool if that could be used to (selectively) share one's library!

**Reply by arunoruto** [reply]
I was more thinking about a way of colaborating, and not only sharing. I am currently using [I, Librarian](https://i-librarian.net/) for our research group, but somehow it lacks something (they have a business model in the background, so they are more focused on that, than on the open source option).

Maybe I will try the nextcloud setup to share the file structure. Syncthing is amazing and I use it for my private servers, but it can be a bit complicated to setup for non-techy persons (I had some trouble with a colleague).

When I am done with writing papers, I will see if I can implement an extension, since I am familiar with python.

**Comment by alexfikl** [comment]
To add to that what @jghauser said:
* The database layer (which is mostly just for caching / querying) in papis is a plugin, so you can teach it to store information somewhere fancier (like S3?).
* The `papis serve` command takes arguments, i.e. `papis serve --port PORT --address ADDRESS`, so if you have a server around, you can run it there and expose the port and it should work.

I mostly run things locally for myself, so I'm not familiar with either of these either, unfortunately. It's probably not a huge amount of effort to make it work, but I don't think your proposed workflow will just work out of the box. We're definitely interested in having some functionality like this though!
